New mod content installer

pchote opened the pull request:
The original content installer has been hacked together over the years, and had a whole bunch of limitations.  This PR completely rewrites it, bringing the following benefits:
- Now understands the difference between different content sources, making it easy to add new ones in the future (Origin / cncnet / GOG etc, as well as translated versions).
- Can now install disk-specific assets (FMVs, expansion music)
- Improved discovery for music installation
- Includes mod-specific description text, making it clear that our asset downloads come from the freeware TD/RA/TS releases
- Removes the modchooser dependency on game-specific package parsers (Installshield, Mix, etc)

There are currently two minor regressions which I would like to leave to followup PRs (the PR is big enough as it is, and these are separate to the main changes):
- We lose TFD support until somebody who has access to the disk can set up the new metadata for it
- The music installation prompt is not shown for players who have the hacked scores.mix from the original assets download.  A future PR will add metadata that tells the game to delete that file as a migration step

Closes #7412, Closes #11344, Closes #10556
Related to #9479, #5970, #8993, #4361, #9082

![screen shot 2016-05-29 at 19 55 37](https://cloud.githubusercontent.com/assets/167819/15691759/8951f17e-2782-11e6-95b8-9f701769b62a.png)
![screen shot 2016-05-29 at 19 56 03](https://cloud.githubusercontent.com/assets/167819/15635515/c729959e-25d7-11e6-854f-779485d12792.png)
![screen shot 2016-05-29 at 19 56 07](https://cloud.githubusercontent.com/assets/167819/15635514/c71cc4b8-25d7-11e6-8cd4-3de7bdf7cff7.png)
![screen shot 2016-05-29 at 19 56 32](https://cloud.githubusercontent.com/assets/167819/15635516/c7354524-25d7-11e6-8f1c-5750dcb7f465.png)

pchote committed:
Add new content installer framework.

pchote committed:
Add command to dump InstallShield contents.

pchote committed:
Add BecameVisible and BecameHidden methods to ChromeLogic.

pchote committed:
Add tooltip support to ImageWidget.

pchote committed:
Allow custom BorderWidths on ScrollPanelWidget.

pchote committed:
Add new content installer metadata to TD.

pchote committed:
Remove legacy music installer.

pchote committed:
Remove legacy installation logic.

pchote committed:
Formatting fixes.

pchote committed:
Add new content installer metadata to D2k.

pchote committed:
Add new content installer metadata to RA.

pchote committed:
Add command to dump Mix contents.

jaZzKCS commented:
Very awesome work.
:+1:

Mailaender commented on OpenRA.Mods.Common/OpenRA.Mods.Common.csproj:
```diff
@@ -748,6 +748,12 @@
+    <Compile Include="Widgets\Logic\Installation\DownloadPackageLogic.cs" />
+    <Compile Include="Widgets\Logic\Installation\ModContentDiscTooltipLogic.cs" />
+    <Compile Include="Widgets\Logic\Installation\InstallFromDiscLogic.cs" />
+    <Compile Include="Widgets\Logic\Installation\ModContentPromptLogic.cs" />
```
These file additions are unrelated to this commit.

pchote commented:
I think this should gracefully handle an empty mirror list and show a "Quick download is not available, please select Advanced install to copy assets from an original game disk" dialog.  This will avoid breakage if we decide/need to remove our online asset mirrors.

pchote committed:
Add new content installer metadata to TS.

pchote commented:
Updated:

* Fixed csproj additions
* Fixed the margins and improved the wording of the asset installation prompt:
![screen shot 2016-05-31 at 22 52 05](https://cloud.githubusercontent.com/assets/167819/15691759/8951f17e-2782-11e6-95b8-9f701769b62a.png)
* Added a graceful fallback if no mirrors are available:
![screen shot 2016-05-31 at 22 50 49](https://cloud.githubusercontent.com/assets/167819/15691793/b445be60-2782-11e6-9c1a-cc09462e7f8c.png)
* Added truncation for long status labels (fixes #10556)

pchote commented:
This uses Assets/Content inconsistently through the UI.  We should standardize on one or the other.

My preference is probably Content.  Does anyone else have an opinion?

Mailaender commented:
Feedback from social media doesn't indicate that it ever resulted in confusion or misunderstandings, so you are probably just fixing your personal urge to unify the wordings, but that is fine as well.

pchote commented:
Updated.

obrakmann commented on OpenRA.Mods.Common/Widgets/Logic/Installation/DownloadPackageLogic.cs:
```diff
@@ -0,0 +1,224 @@
+					return;
+				}
+
+				// Automatically extract
```
What if I don't want the download to be extracted, but just copied straight to Contents/mymod/? Is that an unreasonable use case? Afaik, we support .zip as archive for game files, and I doubt most people would go to the trouble of creating .mix files.

GraionDilach replied:
It's not afaik, I have successfully used zip files in place of mixes and I still plan to do that again.

pchote replied:
I plan on doing exactly this – remember our earlier discussion about defining metadata to download byte ranges directly from a hosted ISO or installer exe?

That's another big job, so I'd prefer to leave this at feature parity with the current code for now.

obrakmann commented:
Code looks good so far. Haven't actually installed anything with it yet, will do that tomorrow.

obrakmann commented:
Also, do you feel like writing a little blurb for https://github.com/OpenRA/OpenRA/wiki/Mod-manifest? :smile:

obrakmann commented on OpenRA.Mods.Common/Widgets/Logic/Installation/DownloadPackageLogic.cs:
```diff
@@ -0,0 +1,224 @@
+					}
+
+					onError("Invalid archive");
+				}
```
Please add a `finally` here that deletes the temp file.

obrakmann commented:
Works very well, good job! :+1: once the temp file is properly cleaned up.

pchote commented:
Added the temp file deletion.

Mailaender commented:
Confirmation that the Dune 2000 video installation works on Windows from https://www.reddit.com/r/openra/comments/4n0rdw/how_to_play_dune2k_cutscenes/d44vfli

Mailaender commented:
It currently crashes

```
Exception of type `System.Reflection.TargetInvocationException`: Exception has been thrown by the target of an invocation.
Inner Exception of type `System.ArgumentNullException`: Value cannot be null.
Parameter name: source
    at System.Linq.Enumerable.Select[TSource,TResult] (IEnumerable`1 source, System.Func`2 selector) [0x0006c] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/external/referencesource/System.Core/System/Linq/Enumerable.cs:44 
  at OpenRA.Mods.Common.Widgets.Logic.ModContentLogic.PopulateContentList () <0x40579520 + 0x004bb> in <filename unknown>:0 
  at OpenRA.Mods.Common.Widgets.Logic.ModContentLogic..ctor (OpenRA.Widgets.Widget widget, System.String modId, System.Action onCancel) <0x40578a90 + 0x0088b> in <filename unknown>:0 
  at (wrapper managed-to-native) System.Reflection.MonoCMethod:InternalInvoke (System.Reflection.MonoCMethod,object,object[],System.Exception&)
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00002] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/mcs/class/corlib/System.Reflection/MonoMethod.cs:644 
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00013] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/mcs/class/corlib/System.Reflection/MonoMethod.cs:650 
  at System.Reflection.MonoCMethod.DoInvoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00089] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/mcs/class/corlib/System.Reflection/MonoMethod.cs:635 
  at System.Reflection.MonoCMethod.Invoke (BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/mcs/class/corlib/System.Reflection/MonoMethod.cs:663 
  at System.Reflection.ConstructorInfo.Invoke (System.Object[] parameters) [0x00000] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/mcs/class/corlib/System.Reflection/ConstructorInfo.cs:62 
  at OpenRA.ObjectCreator.CreateUsingArgs (System.Reflection.ConstructorInfo ctor, System.Collections.Generic.Dictionary`2 args) [0x00062] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/ObjectCreator.cs:140 
  at OpenRA.ObjectCreator.CreateObject[T] (System.String className, System.Collections.Generic.Dictionary`2 args) [0x00063] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/ObjectCreator.cs:105 
  at OpenRA.Widgets.Widget+<PostInit>c__AnonStorey0.<>m__0 (System.String l) [0x00011] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/Widget.cs:274 
  at System.Linq.Enumerable+WhereSelectArrayIterator`2[TSource,TResult].MoveNext () [0x0004d] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/external/referencesource/System.Core/System/Linq/Enumerable.cs:327 
  at System.Linq.Buffer`1[TElement]..ctor (IEnumerable`1 source) [0x00083] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/external/referencesource/System.Core/System/Linq/Enumerable.cs:2568 
  at System.Linq.Enumerable.ToArray[TSource] (IEnumerable`1 source) [0x00011] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/external/referencesource/System.Core/System/Linq/Enumerable.cs:830 
  at OpenRA.Widgets.Widget.PostInit (OpenRA.Widgets.WidgetArgs args) [0x0004c] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/Widget.cs:275 
  at OpenRA.WidgetLoader.LoadWidget (OpenRA.Widgets.WidgetArgs args, OpenRA.Widgets.Widget parent, OpenRA.MiniYamlNode node) [0x001ea] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:75 
  at OpenRA.WidgetLoader.LoadWidget (OpenRA.Widgets.WidgetArgs args, OpenRA.Widgets.Widget parent, System.String w) [0x00032] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:44 
  at OpenRA.Widgets.Ui.OpenWindow (System.String id, OpenRA.Widgets.WidgetArgs args) [0x00012] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/Widget.cs:62 
  at OpenRA.Mods.Common.Widgets.Logic.ModBrowserLogic+<ModBrowserLogic>c__AnonStorey0.<>m__3 () <0x40578910 + 0x00107> in <filename unknown>:0 
  at OpenRA.Mods.Common.Widgets.ButtonWidget.<ButtonWidget>m__A (MouseInput _) <0x40562990 + 0x0001f> in <filename unknown>:0 
  at OpenRA.Mods.Common.Widgets.ButtonWidget.HandleMouseInput (MouseInput mi) <0x4055fed0 + 0x0023d> in <filename unknown>:0 
  at OpenRA.Widgets.Widget.HandleMouseInputOuter (MouseInput mi) [0x000ce] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/Widget.cs:393 
  at OpenRA.Widgets.Ui.HandleInput (MouseInput mi) [0x0002c] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/Widget.cs:102 
  at OpenRA.DefaultInputHandler+<OnMouseInput>c__AnonStorey2.<>m__0 () [0x00006] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Input/InputHandler.cs:50 
  at OpenRA.Sync.CheckSyncUnchanged[T] (OpenRA.World world, System.Func`1 fn) [0x00008] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Sync.cs:174 
  at OpenRA.DefaultInputHandler.OnMouseInput (MouseInput input) [0x00020] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Input/InputHandler.cs:50 
  at OpenRA.Platforms.Default.Sdl2Input.PumpInput (IInputHandler inputHandler) [0x001e7] in /home/matthias/Entwicklung/OpenRA/OpenRA.Platforms.Default/Sdl2Input.cs:107 
  at OpenRA.Platforms.Default.Sdl2GraphicsDevice.PumpInput (IInputHandler inputHandler) [0x0000e] in /home/matthias/Entwicklung/OpenRA/OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs:365 
  at OpenRA.Renderer.EndFrame (IInputHandler inputHandler) [0x0000e] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Renderer.cs:170 
  at OpenRA.Game.RenderTick () [0x00142] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:586 
  at OpenRA.Game.Loop () [0x0013e] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:696 
  at OpenRA.Game.Run () [0x00042] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:714 
  at OpenRA.Program.Run (System.String[] args) [0x00011] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:118 
  at OpenRA.Program.Main (System.String[] args) [0x00050] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:40 
```

when you don't define sources for automatic installation:

```yaml
ModContent:
	InstallPromptMessage: Red Alert 2 requires artwork and audio from the original game.
	HeaderMessage: At the moment only a manual installation is supported.\n\nPlease refer to https://github.com/OpenRA/ra2/wiki#content
	Packages:
		base: Base Game Files
			TestFiles: ^Content/ra2/ra2.mix, ^Content/ra2/language.mix, ^Content/ra2/multi.mix
			Required: true
		music: Base Game Music
			TestFiles: ^Content/ra2/theme.mix
```

pchote committed:
Add a lightweight MSCab parser.

pchote commented:
@Mailaender: That happened because you didn't define any discs.  I fixed your crash, and also left you a present in https://github.com/OpenRA/ra2/pull/233.

The new mscab parser means that @obrakmann should probably recheck at least that code.

Mailaender commented on mods/d2k/mod.yaml:
```diff
@@ -215,3 +199,1986 @@ GameSpeeds:
+				music/ambush.aud: bd5926a56a83bc0e49f96037e1f909014ac0772a
+				setup/setup.z: 937f5ceb1236bf3f3d4e43929305ffe5004078e7
+			Install:
+				copy: MOVIES
```
It is lower case `movies` here for whatever reason.

```
Copying /run/media/matthias/DUNE2000/MOVIES/a_br01_e.vqa -> /home/matthias/.openra/Content/d2k/Movies/A_BR01_E.VQA
System.IO.FileNotFoundException: /run/media/matthias/DUNE2000/MOVIES/a_br01_e.vqa does not exist
File name: '/run/media/matthias/DUNE2000/MOVIES/a_br01_e.vqa'
  at System.IO.File.Copy (System.String sourceFileName, System.String destFileName, Boolean overwrite) [0x001bd] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/mcs/class/corlib/System.IO/File.cs:112 
  at System.IO.File.Copy (System.String sourceFileName, System.String destFileName) [0x00000] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/mcs/class/corlib/System.IO/File.cs:72 
  at OpenRA.Mods.Common.Widgets.Logic.InstallFromDiscLogic+<InstallFromDisc>c__AnonStorey5.<>m__1 () <0x40fea6a0 + 0x003ef> in <filename unknown>:0 
Deleting /home/matthias/.openra/Content/d2k/Movies/A_BR01_E.VQA
```

![image](https://cloud.githubusercontent.com/assets/756669/16054604/297d7a7e-326e-11e6-9db3-350ffba90e6e.png)

Mailaender commented on mods/d2k/mod.yaml:
```diff
@@ -215,3 +199,1986 @@ GameSpeeds:
+					^Content/d2k/Movies/G_PLN2_E.VQA: g_pln2_e.vqa
+					^Content/d2k/Movies/G_PLNT_E.VQA: g_plnt_e.vqa
+					^Content/d2k/Movies/T_TITL_E.VQA: t_titl_e.vqa
+				copy: MUSIC
```
Lower case `music`.

Mailaender commented:
Works :white_check_mark: 

![image](https://cloud.githubusercontent.com/assets/756669/16054120/16052e94-326c-11e6-841c-2c8f24461848.png)

Mailaender commented:
Might be worth putting those lengthy definitions in their own `content.yaml` and just reference a list of MiniYaml definitions like we do for most of the rest.

pchote commented:
> Might be worth putting those lengthy definitions in their own content.yaml and just reference a list of MiniYaml definitions like we do for most of the rest.

I agree in principle, but in practice this is tricky because this metadata is used by the modchooser, and loading cross-mod assets is a bit tricky.

pchote commented:
Fixed d2k linux metadata.  Can you please retest?

Mailaender commented:
Confirmed :white_check_mark: Dune 2000 CD installation is fixed now.

obrakmann commented:
Still looks good to me. Can't really test the ms-cab extraction, though, the only .cabs I found here are Installshield ones (with the ISc* format header)

Still, :+1:

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/51bce65fee22e55479ccedfcf0124d6cf68d0d39)

Mailaender commented:
> We lose TFD support until somebody who has access to the disk can set up the new metadata for it

I generated

https://gist.github.com/Mailaender/69c2410b0ee87ab50513f0dfc73692a0

using

```csharp
#region Copyright & License Information
/*
 * Copyright 2007-2016 The OpenRA Developers (see AUTHORS)
 * This file is part of OpenRA, which is free software. It is made
 * available to you under the terms of the GNU General Public License
 * as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version. For more
 * information, see COPYING.
 */
#endregion

using System;
using System.IO;
using OpenRA.FileSystem;

namespace OpenRA.Mods.Common.UtilityCommands
{
	class ListInstallShieldCabContentsCommand : IUtilityCommand
	{
		public string Name { get { return "--list-hdr"; } }

		public bool ValidateArguments(string[] args)
		{
			return args.Length == 2;
		}

		[Desc("DATA.HDR", "Lists the filenames contained within an InstallShield HDR index file")]
		public void Run(ModData modData, string[] args)
		{
			var filename = Path.GetFileName(args[1]);
			var path = Path.GetDirectoryName(args[1]);

			var fs = new OpenRA.FileSystem.FileSystem();
			fs.Mount(path, "parent");
			var package = new InstallShieldCABExtractor(fs, "parent|" + filename);

			foreach (var file in package.FileDescriptors)
			{
				Console.WriteLine("{0}:", file.Value.Filename);
				Console.WriteLine("\tOffset: {0}", file.Value.DataOffset);
				Console.WriteLine("\tLength: {0}", file.Value.NameOffset);
			}
		}
	}
}
```

MortovMolotov commented:
Hey guys,

I own a TFD disc; how can I help? What do you mean by metadata in this context?

Mailaender commented:
You can help test https://github.com/OpenRA/OpenRA/pull/11536.

Stealthx3 commented:
Is this a firstrun-only prompt like the current one? It would greatly benefit from main menu/settings menu access in the event the user would prefer to add the extra content later.

GraionDilach commented:
No, it isn't. The Manage Content is always available next to the Play button in the mod selection menu.

jtyle6 commented:
Example:
![screen shot 2016-08-13 at 7 25 52 pm](https://cloud.githubusercontent.com/assets/12602985/17642474/17af80c4-618c-11e6-9cbd-813bea25dd21.png)

Stealthx3 commented:
I don't see it, if you are talking about the "Asset browser" (which would be an appropriate place to put it) or the fact that you can install music specifically from the music menu under extras in-game, then both are missing full functionality (the ability to add movies too) and the first-run import/asset grabber does not pull the videos.

![screenshot from 2016-08-15 16-02-51](https://cloud.githubusercontent.com/assets/6005279/17682679/ea70b638-6301-11e6-920c-e81a2a3d5757.png)

GraionDilach commented:
There was no playtest or release since this got merged.

Stealthx3 commented:
Oh ok, that's a relief. The fact that imports can only happen on the first run of any mod rubbed me the wrong way a bit lol.
